src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+    return 0;
+  }
+
+  return 0;
+}
+
+/**
+ * Stages:
+ * PARALLEL
+ *  1. Size and mtime recovery: scan ALL objects, and update 0th
+ *   objects with max size and max mtime seen.
+ * PARALLEL
+ *  2. Inode recovery: scan ONLY 0th objects, and inject metadata
+ *   into dirfrag OMAPs, creating blank dirfrags as needed.  No stats
+ *   or rstats at this stage.  Inodes without backtraces go into
+ *   lost+found
+ * SERIAL
+ *  3. Dirfrag statistics: depth first traverse into metadata tree,
+ *    rebuilding dir sizes.
+ *    XXX hmm, *or* we could use an object class to accumulate
+ *    the number of insertions we've done on any given frag, and
+ *    then do a pass where we just inject that accumulated number
+ *    of insertions into the directory's size and the dirfrags stats?
+ * PARALLEL
+ *  4. Cleanup; go over all 0th objects (and dirfrags if we tagged
+ *   anything onto them) and remove any of the xattrs that we
+ *   used for accumulating.
+ */
+
+
+int parse_oid(const std::string &oid, uint64_t *inode_no, uint64_t *obj_id)
+{
+  if (oid.find(".") == std::string::npos || oid.find(".") == oid.size() - 1) {
+    return -EINVAL;
+  }
+
+  std::string err;
+  std::string inode_str = oid.substr(0, oid.find("."));
+  *inode_no = strict_strtoll(inode_str.c_str(), 16, &err);
+  if (!err.empty()) {
+    return -EINVAL;
+  }
+
+  std::string pos_string = oid.substr(oid.find(".") + 1);
+  *obj_id = strict_strtoll(pos_string.c_str(), 16, &err);
+  if (!err.empty()) {
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+// Pending sharded pgls & add in progress mechanism for that
+#undef SHARDEDPGLS
+
+int DataScan::recover_extents()
+{
+#ifdef SHARDED_PGLS
+  float progress = 0.0;
+  librados::NObjectIterator i = data_io.nobjects_begin(n, m);
+#else
+  librados::NObjectIterator i = data_io.nobjects_begin();
+#endif
+
+  librados::NObjectIterator i_end = data_io.nobjects_end();
+  int r = 0;
+
+  for (; i != i_end; ++i) {
+    const std::string oid = i->get_oid();
+#ifdef SHARDED_PGLS
+    if (i.get_progress() != progress) {
+      if (int(i.get_progress() * 100) / 5 != int(progress * 100) / 5) {
+        std::cerr << percentify(i.get_progress()) << "%" << std::endl;
+      }
+      progress = i.get_progress();
+    }
+#endif
+
+    // Read size
+    uint64_t size;
+    time_t mtime;
+    // FIXME: by using librados, I throw away precision because it wants
+    // to make me use time_t instead of utime_t :-(
+    r = data_io.stat(oid, &size, &mtime);
+    if (r != 0) {
+      dout(4) << "Cannot stat '" << oid << "': skipping" << dendl;
+      continue;
+    }
+
+    // I need to keep track of
+    //  * The highest object ID seen
+    //  * The size of the highest object ID seen
+    //  * The largest object seen
+    //
+    //  Given those things, I can later infer the object chunking
+    //  size, the offset of the last object (chunk size * highest ID seen)
+    //  and the actual size (offset of last object + size of highest ID seen)
```

Comment:
You don't seem to be saving the largest object seen anywhere? It's input to the class, but the class just saves the highest-offset object and its size.